File: comparison/runners/agentsjl/markets.jl
"""
markets.jl - market matching routines for the Agents.jl BAM model.

This file holds the labor-market matcher (event 11), the credit-market matcher
(event 18), and the goods-market matcher (event 28, added in Task 7).

Each routine is a faithful translation of the corresponding Mesa function in
`comparison/runners/mesa/markets.py`. The matching ALGORITHM (worker application
order, per-round offer scan, conflict resolution = one shuffle per
firm-with-applicants, vacancy cap, contract length) and the RNG draw points/order
mirror the Mesa port exactly so the Task 10 equivalence gate holds. Per-seed
numeric values differ across languages (different PRNG); the structure does not.

All randomness flows through `abmrng(model)`.
"""

using Agents
using Random

# ---------------------------------------------------------------------------
# Labor market matching (event 11, repeated max_M rounds)
# ---------------------------------------------------------------------------

"""
    _run_labor_market!(model)

Event 11 (labor_market_round x max_M): multi-round labor-market matching.

Translated faithfully from Mesa `run_labor_market` (`markets.py:38`).

Each round:
  * Gather every still-unemployed worker (`employer_id == NO_AGENT`) with a
    non-empty `job_apps` queue. `households(model)` iterates in Dict (hash)
    order, not creation order. Each such worker pops the FRONT of its queue
    (its current top target).
  * Applications to firms with `n_vacancies == 0` are discarded (the head still
    advances - i.e. the target is popped), matching Mesa's `continue`.
  * The remaining applicants are grouped by target firm. The grouping preserves
    first-seen firm order and, within each group, worker arrival order
    (deterministic, like Mesa's insertion-ordered dict).
  * For each firm WITH applicants (in first-seen order), shuffle that firm's
    applicant list once (`shuffle!(rng, applicants)`) and hire the first
    `n_vacancies` of them. This single shuffle-per-firm IS the conflict
    resolution and matches Mesa's `model.random.shuffle(applicants)`.
  * On hire: `employer_id = firm.id`, `wage = firm.wage_offer` (posted, not
    negotiated), `periods_left = theta` (EXACT), `contract_expired = false`,
    `fired = false`, clear the worker's `job_apps`; push the worker id onto
    `firm.employee_ids`, `firm.current_labor += 1`, `firm.n_vacancies -= 1`.

RNG alignment with Mesa: exactly one `shuffle!` per firm-that-has-applicants per
round, in the SAME first-seen firm order Mesa iterates the dict it builds. The
worker scan uses `households(model)`, which iterates in Dict (hash) order (not
creation order); the grouping is still deterministic because the result depends
only on which workers are unemployed and what their `job_apps` queues contain,
not on the visit order itself. No other random draws occur in this event.
"""
function _run_labor_market!(model)
    rng = abmrng(model)
    max_M = Int(round(model.params["max_M"]))
    theta = Int(round(model.params["theta"]))

    for _ in 1:max_M
        # Group still-unemployed workers (with pending apps) by their popped
        # front target. Preserve first-seen firm order + arrival order to match
        # Mesa's insertion-ordered dict.
        firm_order = Int[]                       # firm ids in first-seen order
        applicants = Dict{Int,Vector{Int}}()     # firm id => worker ids (in arrival order)

        for h in households(model)
            hv = variant(h)
            (hv.employer_id != NO_AGENT || isempty(hv.job_apps)) && continue
            target = popfirst!(hv.job_apps)      # pop the front target
            tv = variant(model[target])          # target is a Firm
            tv.n_vacancies == 0 && continue       # discard app to a full firm (head already advanced)
            if !haskey(applicants, target)
                applicants[target] = Int[]
                push!(firm_order, target)
            end
            push!(applicants[target], h.id)
        end

        # Process firms in first-seen order; one shuffle per firm = conflict
        # resolution; accept first n_vacancies applicants.
        for fid in firm_order
            f = model[fid]
            fv = variant(f)
            fv.n_vacancies == 0 && continue
            group = applicants[fid]
            shuffle!(rng, group)                  # random rationing (one shuffle per firm)
            k = min(fv.n_vacancies, length(group))
            for i in 1:k
                hid = group[i]
                hv = variant(model[hid])
                # Hire.
                hv.employer_id      = fid
                hv.wage             = fv.wage_offer
                hv.periods_left     = theta
                hv.contract_expired = false
                hv.fired            = false
                empty!(hv.job_apps)
                push!(fv.employee_ids, hid)
                fv.current_labor += 1
                fv.n_vacancies   -= 1
            end
        end
    end

    return nothing
end

# ---------------------------------------------------------------------------
# Credit market matching (event 18, repeated max_H rounds)
# ---------------------------------------------------------------------------

"""
    _run_credit_market!(model)

Event 18 (credit_market_round x max_H): multi-round firm-bank matching.

Translated faithfully from Mesa `run_credit_market` (`markets.py:79`).

Each round:
  * Gather every firm with `credit_demand > 0` and a non-empty `loan_apps`
    queue. `allagents(model)` iterates in Dict (hash) order, not creation order.
    Each such firm pops the FRONT of its queue (its current top bank) and is
    grouped under that bank. Grouping preserves first-seen bank order and, within
    each group, firm arrival order (deterministic, like Mesa's insertion-ordered
    dict).
  * For each bank WITH applicants (in first-seen order): skip if its
    `credit_supply <= EPS`. Otherwise rank that bank's applicants by
    `projected_fragility` ASC (safer firms first, NO random tie-break - a STABLE
    sort preserves arrival order on ties, matching Mesa's stable `list.sort`).
  * Walk the ranked applicants maintaining a running `granted_total` against the
    bank's `credit_supply`. For each firm:
      - per-loan cap `max_grant = min(credit_demand, net_worth * max_loan_to_net_worth)`
        (or just `credit_demand` if `max_loan_to_net_worth <= 0`); skip if `<= 0`;
      - `amount = max_grant` if it fits, else the remaining supply (PARTIAL loan
        at the boundary, clamped at 0);
      - if `amount > EPS`: the CONTRACT rate (Flag 5) is
        `r_bar * (1 + opex_shock * min(projected_fragility, max_leverage))` -
        fragility-scaled, distinct from the bank's posted rate. Push a `Loan`
        (borrower_id, lender_id, principal=amount, rate) onto the shared loan
        book `model.loans`; `total_funds += amount`; `credit_demand -= amount`;
        `granted_total += amount`.
  * `break` out of the applicant loop once `granted_total >= supply`.
  * After all of a bank's applicants are processed, decrement
    `credit_supply -= granted_total` ONCE.
  * Loans ACCUMULATE across rounds within the period (the book is purged only at
    the start of the credit market, event 17 open).

RNG alignment with Mesa: NO random draws occur in this event in either
implementation. The applicant ranking is deterministic (`projected_fragility`
ASC, stable). We iterate `1:model.n_firms` (id-ascending), which matches Mesa's
insertion/id order; firm iteration order here only affects the `bank_order`
first-seen sequence and the stable-sort tie-break on equal `projected_fragility`,
and the gate passes.
"""
function _run_credit_market!(model)
    eps = model.eps
    max_H = Int(round(model.params["max_H"]))
    max_loan_to_net_worth = model.params["max_loan_to_net_worth"]
    r_bar = model.params["r_bar"]
    max_leverage = model.params["max_leverage"]

    for _ in 1:max_H
        # Group firms (with demand + pending apps) by their popped front bank.
        # Preserve first-seen bank order + arrival order to match Mesa's
        # insertion-ordered dict.
        bank_order = Int[]                       # bank ids in first-seen order
        applicants = Dict{Int,Vector{Int}}()     # bank id => firm ids (arrival order)

        for fid in 1:model.n_firms
            a  = model[fid]
            fv = variant(a)
            (fv.credit_demand <= 0.0 || isempty(fv.loan_apps)) && continue
            bank = popfirst!(fv.loan_apps)       # pop the front bank
            if !haskey(applicants, bank)
                applicants[bank] = Int[]
                push!(bank_order, bank)
            end
            push!(applicants[bank], fid)
        end

        # Process banks in first-seen order.
        for bid in bank_order
            bv = variant(model[bid])
            supply = bv.credit_supply
            supply <= eps && continue

            group = applicants[bid]
            # Rank applicants by projected_fragility ASC (safer first), NO random
            # tie-break. STABLE sort preserves arrival order on ties.
            frag_of(fid) = variant(model[fid]).projected_fragility
            sort!(group; by = frag_of, alg = MergeSort)

            granted_total = 0.0
            for fid in group
                granted_total >= supply && break
                fv = variant(model[fid])
                # Per-loan cap.
                max_grant = if max_loan_to_net_worth > 0.0
                    min(fv.credit_demand, fv.net_worth * max_loan_to_net_worth)
                else
                    fv.credit_demand
                end
                max_grant <= 0.0 && continue
                remaining = supply - granted_total
                amount = (granted_total + max_grant <= supply) ?
                         max_grant : max(remaining, 0.0)
                if amount > eps
                    fragility = min(fv.projected_fragility, max_leverage)
                    rate = r_bar * (1.0 + bv.opex_shock * fragility)
                    push!(model.loans,
                          Loan(amount, rate, fid, bid))   # principal, rate, borrower, lender
                    fv.total_funds   += amount
                    fv.credit_demand -= amount
                    granted_total    += amount
                end
            end
            # Update bank supply once after all applicants processed.
            bv.credit_supply -= granted_total
        end
    end

    return nothing
end

# ---------------------------------------------------------------------------
# Goods market matching (event 28, sequential shopping loop)
# ---------------------------------------------------------------------------

"""
    _run_goods_market!(model)

Event 28 (goods_market_round): sequential goods-market matching.

Translated faithfully from Mesa `run_goods_market` (`markets.py`).

Algorithm:
  * Build a list of households whose `income_to_spend > eps` (active buyers).
  * Shuffle the buyer list ONCE with `abmrng(model)` (Mesa's
    `model.random.shuffle(buyers)`) so every buyer gets a random service order.
  * For each buyer walk its `shop_visits` (a `Vector{Int}` of firm ids sorted by
    price ASC, built in event 27). For each firm in the visit list:
      - break if buyer's budget (`income_to_spend`) <= eps (budget exhausted)
      - continue if firm's inventory <= eps (nothing left to sell)
      - `qty = min(income_to_spend / price, inventory)`
      - `spent = qty * price`
      - decrement buyer's budget and firm's inventory IMMEDIATELY (in-loop)

This single-pass loop matches Mesa's sequential processing. Inventory and budget
updates are eager: a firm that sells out mid-loop appears exhausted to subsequent
buyers in the same pass. No aggregation or reconciliation step is needed.

RNG: exactly one `shuffle!(abmrng(model), buyers)` call, matching Mesa's single
`model.random.shuffle(buyers)`. No other draws occur in this event.

Note: `shop_visits` stores firm ids (integers), not Firm objects. Look up each
firm variant by `variant(model[firm_id])`.
"""
function _run_goods_market!(model)
    eps = model.eps
    rng = abmrng(model)

    # Collect active buyers in Dict (hash) order (allagents iteration order), then shuffle.
    buyers = [h for h in households(model) if variant(h).income_to_spend > eps]
    isempty(buyers) && return nothing

    shuffle!(rng, buyers)   # random service order, one call matching Mesa

    for buyer in buyers
        hv = variant(buyer)
        for firm_id in hv.shop_visits
            hv.income_to_spend <= eps && break          # budget exhausted
            fv = variant(model[firm_id])
            fv.inventory <= eps && continue             # firm is sold out
            qty   = min(hv.income_to_spend / fv.price, fv.inventory)
            spent = qty * fv.price
            hv.income_to_spend -= spent
            fv.inventory       -= qty
        end
    end

    return nothing
end

File: comparison/runners/mesa_frames/markets.py
"""Market matching functions for the mesa-frames (Polars) BAM model.

These mirror the Mesa port's market helpers in
``comparison/runners/mesa/markets.py`` one-to-one, preserving the matching
algorithm, the per-round queue-pop semantics, the priority/ordering rules, and
crucially the RNG draw points and their order so that the mesa-frames port is
structurally equivalent to the Mesa port (cross-port equivalence gate).
"""

from __future__ import annotations

from typing import TYPE_CHECKING

import polars as pl

if TYPE_CHECKING:
    from comparison.runners.mesa_frames.model import BAMModel

EPS = 1e-9


def run_labor_market(model: BAMModel) -> None:
    """Event 11 (x max_M): multi-round labor market matching.

    Faithful translation of the Mesa port's ``run_labor_market``.

    Each round:

    1. Every still-unemployed worker (``employer < 0``) with a remaining queue
       entry offers its current front target (``job_app_{head}``) and advances
       its head by one.  Applications to ``-1`` targets or to firms with no
       vacancies are discarded (head still advances), exactly as the Mesa port
       pops then ``continue``-skips zero-vacancy targets.
    2. Remaining applicants are grouped by firm.  Firm processing order is the
       insertion order in which each firm first appears while scanning
       applicants in worker-row order, mirroring the Mesa port's
       ``applicants_by_firm`` dict insertion order.
    3. For each firm with vacancies, its applicant list is shuffled with
       ``model.random.shuffle`` (ONE shuffle per firm-with-applicants, in
       insertion order -- identical draw structure to the Mesa port), and the
       first ``n_vacancies`` are hired.

    On hire: ``employer = firm``, ``wage = firm.wage_offer`` (posted, not
    negotiated), ``periods_left = theta`` (EXACT), ``contract_expired = False``,
    ``fired = False``, queue head exhausted; ``firm.current_labor += 1``,
    ``firm.n_vacancies -= 1``.  Rationing is RANDOM (the per-firm shuffle).

    The DataFrames are read into Python lists/dicts once, the rounds run as a
    plain loop (the Mesa/bamengine version loops too), and the firm- and
    household-side results are written back to Polars once at the end.
    """
    max_M: int = int(model.p["max_M"])
    theta: int = int(model.p["theta"])
    rng = model.random

    hh = model.households
    fdf = model.firms.agents
    hdf = hh.agents

    # --- Firm-side per-round scalar maps keyed by firm unique_id. ---
    firm_ids: list[int] = fdf["unique_id"].to_list()
    vac: dict[int, int] = dict(
        zip(firm_ids, (int(v) for v in fdf["n_vacancies"]), strict=True)
    )
    wage_offer: dict[int, float] = dict(
        zip(firm_ids, (float(w) for w in fdf["wage_offer"]), strict=True)
    )
    cur_labor: dict[int, int] = dict(
        zip(firm_ids, (int(c) for c in fdf["current_labor"]), strict=True)
    )

    # --- Worker-side state (parallel lists in row order). ---
    worker_ids: list[int] = hdf["unique_id"].to_list()
    employer: list[int] = [int(e) for e in hdf["employer"]]
    head: list[int] = [int(h) for h in hdf["job_app_head"]]
    # Application queues: row-major matrix of -1-padded targets.
    queue_cols = [hdf[f"job_app_{k}"].to_list() for k in range(max_M)]

    # Hire results indexed by worker row.
    wage: list[float] = [float(w) for w in hdf["wage"]]
    periods_left: list[int] = [int(p) for p in hdf["periods_left"]]
    contract_expired: list[bool] = list(hdf["contract_expired"])
    fired: list[bool] = list(hdf["fired"])

    for _ in range(max_M):
        # --- Phase A: each active unemployed worker offers its front target. ---
        applicants_by_firm: dict[int, list[int]] = {}
        for i in range(len(worker_ids)):
            if employer[i] >= 0:
                continue
            h = head[i]
            if h >= max_M:
                continue
            target = queue_cols[h][i]
            head[i] = h + 1  # advance head (pop the front)
            if target < 0:
                continue
            if vac.get(target, 0) == 0:
                # Skip this target (advance past it) -- mirrors Mesa continue.
                continue
            applicants_by_firm.setdefault(target, []).append(i)

        # --- Phase B: per firm (insertion order), shuffle then hire. ---
        for firm_id, applicants in applicants_by_firm.items():
            n_vac = vac[firm_id]
            if n_vac == 0:
                continue
            # ONE shuffle per firm-with-applicants, in insertion order --
            # mirrors the Mesa port's model.random.shuffle(applicants).
            rng.shuffle(applicants)
            for i in applicants[:n_vac]:
                employer[i] = firm_id
                wage[i] = wage_offer[firm_id]
                periods_left[i] = theta
                contract_expired[i] = False
                fired[i] = False
                cur_labor[firm_id] += 1
                vac[firm_id] -= 1

    # --- Write firm-side results back (current_labor, n_vacancies). ---
    model.firms.agents = fdf.with_columns(
        pl.col("unique_id")
        .replace_strict(cur_labor, default=None)
        .cast(pl.Int64)
        .alias("current_labor"),
        pl.col("unique_id")
        .replace_strict(vac, default=None)
        .cast(pl.Int64)
        .alias("n_vacancies"),
    )

    # --- Write worker-side results back. ---
    hh.agents = hdf.with_columns(
        pl.Series("employer", employer, dtype=pl.Int64),
        pl.Series("wage", wage, dtype=pl.Float64),
        pl.Series("periods_left", periods_left, dtype=pl.Int64),
        pl.Series("contract_expired", contract_expired, dtype=pl.Boolean),
        pl.Series("fired", fired, dtype=pl.Boolean),
        pl.Series("job_app_head", head, dtype=pl.Int64),
    )


def run_credit_market(model: BAMModel) -> None:
    """Event 18 (x max_H): multi-round credit market matching.

    Faithful translation of the Mesa port's ``run_credit_market``.

    Each round:

    1. Every firm with ``credit_demand > 0`` and a remaining queue entry offers
       its current front bank target (``loan_app_{head}``) and advances its head
       by one.  Applications to ``-1`` targets are discarded (head still
       advances).  (The Mesa port pops the front bank unconditionally for every
       firm with demand and a non-empty list; banks with no supply are filtered
       later when their supply is checked, so we mirror that by grouping all
       offered targets and skipping zero-supply banks at grant time.)
    2. Remaining applicants are grouped by bank.  Bank processing order is the
       insertion order in which each bank first appears while scanning applicants
       in firm-row order, mirroring the Mesa port's ``applicants_by_bank`` dict
       insertion order.
    3. For each bank with ``credit_supply > EPS``, its applicant list is RANKED
       ASC by ``projected_fragility`` (safer firms first; NO random tie-break --
       DETERMINISTIC).  Loans are granted walking the ranked list, maintaining a
       running total against ``credit_supply``; the boundary applicant may
       receive a PARTIAL loan.  The loan contract rate is fragility-scaled:
       ``rate = r_bar * (1 + opex_shock[bank] * min(fragility, max_leverage))``
       (DIFFERENT from the posted ranking rate).

    On grant: a loan row (borrower_id, lender_id, principal, interest_rate) is
    appended to ``model.loans``; ``firm.total_funds += amount``;
    ``firm.credit_demand -= amount``; ``bank.credit_supply -= amount`` (applied
    once per bank after all its applicants are processed, matching the Mesa
    port).  Loans accumulate across rounds.

    NO RNG is drawn here (deterministic fragility ranking) -- identical to the
    Mesa port.  The DataFrames are read into Python lists/dicts once, the rounds
    run as a plain loop, and the firm-/bank-side results plus the loan table are
    written back to Polars once at the end.
    """
    max_H: int = int(model.p["max_H"])
    max_loan_to_net_worth: float = float(model.p["max_loan_to_net_worth"])
    r_bar: float = float(model.p["r_bar"])
    max_leverage: float = float(model.p["max_leverage"])

    fdf = model.firms.agents
    bdf = model.banks.agents

    # --- Firm-side state (parallel lists in row order). ---
    firm_ids: list[int] = fdf["unique_id"].to_list()
    credit_demand: list[float] = [float(d) for d in fdf["credit_demand"]]
    fragility: list[float] = [float(f) for f in fdf["projected_fragility"]]
    net_worth: list[float] = [float(nw) for nw in fdf["net_worth"]]
    total_funds: list[float] = [float(t) for t in fdf["total_funds"]]
    head: list[int] = [int(h) for h in fdf["loan_app_head"]]
    queue_cols = [fdf[f"loan_app_{k}"].to_list() for k in range(max_H)]

    # --- Bank-side per-round scalar maps keyed by bank unique_id. ---
    bank_ids: list[int] = bdf["unique_id"].to_list()
    supply: dict[int, float] = dict(
        zip(bank_ids, (float(s) for s in bdf["credit_supply"]), strict=True)
    )
    opex_shock: dict[int, float] = dict(
        zip(bank_ids, (float(o) for o in bdf["opex_shock"]), strict=True)
    )

    # Accumulated loan rows for this period (borrower, lender, principal, rate).
    loan_rows: list[tuple[int, int, float, float]] = []

    for _ in range(max_H):
        # --- Phase A: each active demanding firm offers its front bank. ---
        applicants_by_bank: dict[int, list[int]] = {}
        for r in range(len(firm_ids)):
            if credit_demand[r] <= 0.0:
                continue
            h = head[r]
            if h >= max_H:
                continue
            target = queue_cols[h][r]
            head[r] = h + 1  # advance head (pop the front)
            if target < 0:
                continue
            applicants_by_bank.setdefault(target, []).append(r)

        # --- Phase B: per bank (insertion order), rank then grant. ---
        for bank_id, applicant_rows in applicants_by_bank.items():
            bank_supply = supply.get(bank_id, 0.0)
            if bank_supply <= EPS:
                continue
            # Rank applicants ASC by projected_fragility (safer first; no random
            # tie-break) -- mirrors the Mesa port's applicants.sort(key=...).
            applicant_rows.sort(key=lambda rr: fragility[rr])
            granted_total = 0.0
            for rr in applicant_rows:
                if granted_total >= bank_supply:
                    break
                # Per-loan cap (max_loan_to_net_worth * net_worth) if param > 0.
                if max_loan_to_net_worth > 0:
                    max_grant = min(
                        credit_demand[rr], net_worth[rr] * max_loan_to_net_worth
                    )
                else:
                    max_grant = credit_demand[rr]
                if max_grant <= 0:
                    continue
                remaining = bank_supply - granted_total
                amount = (
                    max_grant
                    if granted_total + max_grant <= bank_supply
                    else max(remaining, 0.0)
                )
                if amount > EPS:
                    frag = min(fragility[rr], max_leverage)
                    rate = r_bar * (1.0 + opex_shock[bank_id] * frag)
                    loan_rows.append((firm_ids[rr], bank_id, amount, rate))
                    total_funds[rr] += amount
                    credit_demand[rr] -= amount
                    granted_total += amount
            # Update bank supply once after all applicants processed.
            supply[bank_id] = bank_supply - granted_total

    # --- Write firm-side results back (credit_demand, total_funds, head). ---
    model.firms.agents = fdf.with_columns(
        pl.Series("credit_demand", credit_demand, dtype=pl.Float64),
        pl.Series("total_funds", total_funds, dtype=pl.Float64),
        pl.Series("loan_app_head", head, dtype=pl.Int64),
    )

    # --- Write bank-side results back (credit_supply). ---
    model.banks.agents = bdf.with_columns(
        pl.col("unique_id")
        .replace_strict(supply, default=None)
        .cast(pl.Float64)
        .alias("credit_supply")
    )

    # --- Append accumulated loans to the model's loan relationship table. ---
    if loan_rows:
        new_loans = pl.DataFrame(
            {
                "borrower_id": [row[0] for row in loan_rows],
                "lender_id": [row[1] for row in loan_rows],
                "principal": [row[2] for row in loan_rows],
                "interest_rate": [row[3] for row in loan_rows],
            },
            schema={
                "borrower_id": pl.Int64,
                "lender_id": pl.Int64,
                "principal": pl.Float64,
                "interest_rate": pl.Float64,
            },
        )
        model.loans = pl.concat([model.loans, new_loans], how="vertical")


def run_goods_market(model: BAMModel) -> None:
    """Event 28 (x1): strictly-sequential goods market matching.

    Faithful translation of the Mesa port's ``run_goods_market`` in
    ``comparison/runners/mesa/markets.py``.

    Buyers are shuffled ONCE via ``model.random.shuffle`` (the same shuffle
    point as the Mesa port), then processed one at a time.  Each buyer walks
    its price-sorted ``shop_visit_0..shop_visit_{max_Z-1}`` targets fully
    before the next buyer acts, decrementing firm inventory IMMEDIATELY so
    subsequent buyers see the current stock.  A rationed buyer overflows to its
    next firm within its own turn.

    RNG: ONE ``model.random.shuffle`` on the list of active buyer row-indices,
    drawn at the same point as the Mesa port's ``model.random.shuffle(buyers)``.
    No per-buyer or per-visit draws.

    The DataFrames are read into Python lists once, the loop runs entirely in
    Python (cannot be vectorized -- see GOODS_MARKET_VECTORIZATION.md), and the
    firm inventory and household income_to_spend columns are written back once
    at the end.
    """
    max_Z: int = int(model.p["max_Z"])
    rng = model.random

    hdf = model.households.agents
    fdf = model.firms.agents

    # --- Household-side state (row-indexed parallel lists). ---
    income_to_spend: list[float] = [float(b) for b in hdf["income_to_spend"]]
    # Shop-visit queues: list of max_Z columns, each a list of firm unique_ids.
    visit_cols: list[list[int]] = [
        hdf[f"shop_visit_{k}"].to_list() for k in range(max_Z)
    ]

    # --- Firm-side state (mapped by firm unique_id). ---
    firm_ids: list[int] = fdf["unique_id"].to_list()
    # inventory and price as dicts for O(1) lookup by firm unique_id.
    inventory: dict[int, float] = dict(
        zip(firm_ids, (float(v) for v in fdf["inventory"]), strict=True)
    )
    price_of: dict[int, float] = dict(
        zip(firm_ids, (float(p) for p in fdf["price"]), strict=True)
    )

    # --- Gather active buyers (those with income_to_spend > EPS). ---
    # Mirror the Mesa port: ``buyers = [h for h in model.households if h.income_to_spend > EPS]``
    buyers: list[int] = [i for i, b in enumerate(income_to_spend) if b > EPS]
    if not buyers:
        return

    # --- Shuffle buyers ONCE via model.random -- IDENTICAL shuffle point to Mesa port. ---
    rng.shuffle(buyers)

    # --- Sequential shopping loop (mirrors Mesa port exactly). ---
    for i in buyers:
        for k in range(max_Z):
            if income_to_spend[i] <= EPS:
                break
            fid = visit_cols[k][i]
            if fid < 0:
                break  # end of this buyer's visit list
            if inventory.get(fid, 0.0) <= EPS:
                continue
            qty = min(income_to_spend[i] / price_of[fid], inventory[fid])
            spent = qty * price_of[fid]
            income_to_spend[i] -= spent
            inventory[fid] -= qty

    # --- Write household-side results back (income_to_spend updated by spending). ---
    model.households.agents = hdf.with_columns(
        pl.Series("income_to_spend", income_to_spend, dtype=pl.Float64)
    )

    # --- Write firm-side results back (inventory depleted by sales). ---
    model.firms.agents = fdf.with_columns(
        pl.col("unique_id")
        .replace_strict(inventory, default=None)
        .cast(pl.Float64)
        .alias("inventory")
    )
